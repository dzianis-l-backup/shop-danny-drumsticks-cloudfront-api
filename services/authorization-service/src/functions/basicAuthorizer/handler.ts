enum HttpStatuses {
    Unauthorized = 401,
    Forbidden = 403,
    Ok = 200,
}

enum Effect {
    Allow = "Allow",
    Deny = "Deny",
}

const validateCredentials = (username: string, password: string) => {
    return (
        username === process.env.AUTH_USERNAME &&
        password === process.env.AUTH_PASSWORD
    )
}

const getPolicy = (
    principalId: string,
    effect: Effect,
    resource: string,
    statusCode,
    message
) => {
    return {
        principalId,
        policyDocument: {
            Version: "2012-10-17",

            Statement: [
                {
                    Action: "execute-api:Invoke",
                    Effect: effect,
                    Resource: [resource],
                },
            ],
        },
        context: {
            statusCode,
            message,
        },
    }
}

const processToken = (authorizationHeader: string) => {
    const responseUnauthorized = {
        Effect: Effect.Deny,
        statusCode: HttpStatuses.Unauthorized,
        message: "Unauthorized ðŸ˜¢",
        token: authorizationHeader,
    }

    const responseAuthorized = {
        Effect: Effect.Allow,
        statusCode: HttpStatuses.Ok,
        message: "Ok",
    }

    if (!authorizationHeader) {
        return responseUnauthorized
    }

    const [, token] = authorizationHeader.split("Basic ")

    if (!token) {
        return responseUnauthorized
    }

    const [username, password] = Buffer.from(token, "base64")
        .toString()
        .split(":")

    if (!username || !password) {
        return { ...responseUnauthorized, token }
    }

    if (!validateCredentials(username, password)) {
        return { ...responseUnauthorized, token: username }
    }

    return { ...responseAuthorized, token: username }
}

export const basicAuthorizer = async (event) => {
    debugger
    console.log("BASIC_AUTHORIZER EVENT", event)
    const authorizationHeader = event.authorizationToken
    console.log("authorizationHeader", authorizationHeader)

    const { token, Effect, statusCode, message } =
        processToken(authorizationHeader)

    const policy = getPolicy(
        token,
        Effect,
        event.methodArn,
        statusCode,
        message
    )

    console.log("policy", policy)

    return policy
}
