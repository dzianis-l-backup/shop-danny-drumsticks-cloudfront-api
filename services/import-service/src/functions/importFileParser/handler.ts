import { HttpStatuses } from "@const/index"
import { logger } from "@libs/logger"
import AWS from "aws-sdk"
import csv from "csv-parser"

export const importFileParser = async (event) => {
    const s3 = new AWS.S3({ region: process.env.REGION })

    try {
        for (const record of event.Records) {
            logger.log("incomming record", record)

            await new Promise((resolve) => {
                s3.getObject({
                    Bucket: process.env.BUCKET_CSV,
                    Key: record.s3.object.key,
                })
                    .createReadStream()
                    .pipe(csv())
                    .on("data", (data) => logger.log(data))
                    .on("end", async () => {
                        const source = `${process.env.BUCKET_CSV}/${record.s3.object.key}`
                        const distKey = record.s3.object.key.replace(
                            process.env.BUCKET_CSV_SOURCE,
                            process.env.BUCKET_CSV_DIST
                        )

                        logger.log(`Copy Source ${source}`)
                        logger.log(
                            `Destination Source ${process.env.BUCKET_CSV}/${distKey}`
                        )

                        await s3
                            .copyObject({
                                Bucket: process.env.BUCKET_CSV,
                                CopySource: source,
                                Key: distKey,
                            })
                            .promise()

                        logger.log(
                            `Copied from ${source} to ${process.env.BUCKET_CSV}/${distKey}`
                        )

                        await s3
                            .deleteObject({
                                Bucket: process.env.BUCKET_CSV,
                                Key: record.s3.object.key,
                            })
                            .promise()

                        logger.log(`Deleted from ${source}`)

                        resolve({
                            statusCode: HttpStatuses.Ok,
                        })
                    })
                    .on("error", (error) => {
                        logger.error(error)

                        resolve({
                            statusCode: HttpStatuses.InternalServerError,
                        })
                    })
            })
        }
    } catch (error) {
        logger.error(error)
    }
}
