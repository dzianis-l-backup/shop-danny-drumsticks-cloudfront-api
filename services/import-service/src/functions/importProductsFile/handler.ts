import { formatJSONResponse } from "@libs/api-gateway"
import { HttpStatuses } from "@const/index"
import AWS from "aws-sdk"

export const importProductsFile = async (event) => {
    try {
        const { name: fileName } = event.queryStringParameters || {}
        const s3 = new AWS.S3({ region: process.env.REGION })

        if (!fileName) {
            return formatJSONResponse({ statusCode: HttpStatuses.BadRequest })
        }

        const dist = `${process.env.BUCKET_CSV_SOURCE}/${fileName}`
        const params = {
            Bucket: process.env.BUCKET_CSV,
            Key: dist,
            Expires: 60,
            ContentType: "text/csv",
        }
        const url = await s3.getSignedUrlPromise("putObject", params)

        console.log(`Î» importProductsFile url`, url)

        return formatJSONResponse({ statusCode: HttpStatuses.Ok, payload: url })
    } catch (error) {
        console.log(error)

        return formatJSONResponse({
            statusCode: HttpStatuses.InternalServerError,
        })
    }
}
