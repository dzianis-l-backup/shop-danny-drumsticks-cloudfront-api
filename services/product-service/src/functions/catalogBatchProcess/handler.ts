import { SQSEvent, Context } from "aws-lambda"
import { SNS } from "aws-sdk"

import { logger } from "@libs/logger"
import {
    ControllerResponse,
    HttpStatuses,
    Stick,
    StickStockRaw,
    Stock,
    BroadCasts,
} from "src/types"
import { ProductsController } from "@controllers/productsController"

const refineRecordBody = (record: any) => {
    const body = JSON.parse(record.body)

    return {
        title: body.title,
        description: body.description,
        price: Number(body.price),
        count: Number(body.count),
    }
}

const priceThreshold = 1

export const catalogBatchProcess = async (
    event: SQSEvent,
    context: Context
) => {
    logger.log("event", event)
    logger.log("context", context)

    const sns = new SNS({ region: process.env.REGION })
    const records = event.Records.map(refineRecordBody)
    const response: ControllerResponse<ControllerResponse<[Stick, Stock]>[]> =
        await ProductsController.catalogBatchProcess(records as StickStockRaw[])

    try {
        await sns
            .publish({
                Subject: "Hey, here is the outcome to batch catalog creation",
                Message: JSON.stringify(response),
                MessageAttributes: {
                    broadcast: {
                        DataType: "String",
                        StringValue: BroadCasts.Business,
                    },
                },
                TopicArn: process.env.SNS_TOPIC_CREATE_BATCH_PROCESS_ARN,
            })
            .promise()

        if (
            response.payload.some((p) => p.payload[0]?.price <= priceThreshold)
        ) {
            await sns
                .publish({
                    Subject: "Hey, there are some cheap sticks, hurry to buy",
                    Message: JSON.stringify(response),
                    MessageAttributes: {
                        broadcast: {
                            DataType: "String",
                            StringValue: BroadCasts.Private,
                        },
                    },
                    TopicArn: process.env.SNS_TOPIC_CREATE_BATCH_PROCESS_ARN,
                })
                .promise()
        }

        return response
    } catch (error) {
        logger.error(error)
        return { statusCode: HttpStatuses.INTERNAL_SERVER_ERROR }
    }
}
