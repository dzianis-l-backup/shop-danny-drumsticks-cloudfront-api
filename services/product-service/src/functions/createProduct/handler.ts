import { formatJSONResponse } from "@libs/api-gateway"
import { ProductsController } from "@controllers/productsController"
import { StickStock } from "src/types"
import { middyfyPost } from "@libs/lambda"

export const createProductHandler = async ({
    body,
}: {
    body: Omit<StickStock, StickStock["id"]>
}) => {
    const { payload: stick, statusCode } =
        await ProductsController.createProduct(body)

    return formatJSONResponse({ payload: stick, statusCode })
}

export const createProduct = middyfyPost(createProductHandler)
