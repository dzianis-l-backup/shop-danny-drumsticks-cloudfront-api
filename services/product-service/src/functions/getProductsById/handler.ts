import { formatJSONResponse } from "@libs/api-gateway"
import { ProductsController } from "@controllers/productsController"

export const getProductsById = async (event) => {
    const id = event?.pathParameters?.id
    const [stick, error] = await ProductsController.getProductsById(id)

    return formatJSONResponse({ payload: stick, error })
}
