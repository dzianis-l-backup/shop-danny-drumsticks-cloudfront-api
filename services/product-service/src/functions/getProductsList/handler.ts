import { formatJSONResponse } from "@libs/api-gateway"
import { ProductsController } from "@controllers/productsController"

export const getProductsList = async () => {
    const { payload, statusCode } = await ProductsController.getProductsList()

    return formatJSONResponse({ payload, statusCode })
}
