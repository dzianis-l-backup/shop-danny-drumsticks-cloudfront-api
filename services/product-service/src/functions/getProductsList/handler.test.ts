import { getProductsListHandler } from "./handler"

describe("product-service", () => {
    describe("getProductsListHandler", () => {
        it("should return the list of available products", async () => {
            const response = await getProductsListHandler()

            expect(JSON.parse(response.body)).toMatchSnapshot()
        })
    })
})
