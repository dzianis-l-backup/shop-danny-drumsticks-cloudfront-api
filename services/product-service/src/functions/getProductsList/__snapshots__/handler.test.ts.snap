// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`product-service getProductsListHandler should return the list of available products 1`] = `
[
  {
    "count": 1,
    "description": "Balancing power, finesse, and durability, the 5B stick size is a longstanding rock favorite.",
    "id": "aaaaaaaa-aaaa-aaaa-aaaa-aaaaaaaaaaaa",
    "price": 1,
    "title": "vater5b",
  },
  {
    "count": 1,
    "description": "Vic Firth 5A American Classic Hickory.",
    "id": "bbbbbbbb-bbbb-bbbb-bbbb-bbbbbbbbbbbb",
    "price": 1,
    "title": "vicfirth5a",
  },
  {
    "count": 1,
    "description": "Vic Firth 5B American classic drumsticks line combines tradition and Vic Firth style.",
    "id": "cccccccc-cccc-cccc-cccc-cccccccccccc",
    "price": 1,
    "title": "vicfirth5b",
  },
]
`;
